File: webflux-dialogue/src/main/java/com/study/webflux/rag/domain/dialogue/model/ConversationSession.java
package com.study.webflux.rag.domain.dialogue.model;

import java.time.Instant;

public record ConversationSession(
	ConversationSessionId sessionId,
	PersonaId personaId,
	UserId userId,
	Instant createdAt,
	Instant deletedAt
) {
	public ConversationSession {
		if (sessionId == null) {
			throw new IllegalArgumentException("sessionId cannot be null");
		}
		if (personaId == null) {
			personaId = PersonaId.defaultPersona();
		}
		if (userId == null) {
			throw new IllegalArgumentException("userId cannot be null");
		}
		if (createdAt == null) {
			createdAt = Instant.now();
		}
	}

	public static ConversationSession create(PersonaId personaId, UserId userId) {
		return new ConversationSession(
			ConversationSessionId.generate(),
			personaId,
			userId,
			Instant.now(),
			null);
	}

	public boolean isActive() {
		return deletedAt == null;
	}

	public ConversationSession softDelete() {
		return new ConversationSession(sessionId, personaId, userId, createdAt, Instant.now());
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/domain/dialogue/model/ConversationSession.kt
package com.miyou.app.domain.dialogue.model

import java.time.Instant

data class ConversationSession(
    val sessionId: ConversationSessionId,
    val personaId: PersonaId = PersonaId.defaultPersona(),
    val userId: UserId,
    val createdAt: Instant = Instant.now(),
    val deletedAt: Instant? = null,
) {
    init {
        require(personaId.value.isNotBlank()) { "personaId cannot be blank" }
        require(userId.value.isNotBlank()) { "userId cannot be blank" }
    }

    fun isActive(): Boolean = deletedAt == null

    fun softDelete(): ConversationSession = copy(deletedAt = Instant.now())

    companion object {
        @JvmStatic
        fun create(
            personaId: PersonaId,
            userId: UserId,
        ): ConversationSession =
            ConversationSession(
                sessionId = ConversationSessionId.generate(),
                personaId = personaId,
                userId = userId,
            )
    }

    fun sessionId(): ConversationSessionId = sessionId

    fun personaId(): PersonaId = personaId

    fun userId(): UserId = userId

    fun createdAt(): java.time.Instant = createdAt

    fun deletedAt(): java.time.Instant? = deletedAt
}
